src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs

```diff
@@ -959,7 +967,10 @@ protected override void ProcessRecord()
                     try
                     {
                         // Change the default location stack
+                        StackName = ReplacePathWithLastLocationIfApplicable(StackName);
+                        var initialPath = SessionState.Path.CurrentLocation.Path;
                         result = SessionState.Path.SetDefaultLocationStack(StackName);
+                        Environment.SetEnvironmentVariable(environmentVariableNameForPreviousLocation, initialPath);
```

Comment:
The previous comments apply here, including moving the logic to `SessionState.Path.SetDefaultLocationStack`.